### ISSUE TITLE
bibxml-service needs to provide URLs for xi:includes

### ISSUE THREAD

**Issue by ajeanmahoney** [ietf-tools/bibxml-service#225]
A main use case for this service is to get the bibxml URL for a document in order to include it in an RFCXML file. For instance, when searching for an I-D (e.g., draft-ietf-tcpm-rfc793bis), the document page that is returned should have the following URL featured prominently:

    https://bib.ietf.org/public/rfc/bibxml3/reference.I-D.ietf-tcpm-rfc793bis.xml

So that the user can use it in the following XML in the references section:

    <xi:include href="https://bib.ietf.org/public/rfc/bibxml3/reference.I-D.ietf-tcpm-rfc793bis.xml"/>

**Comment by ronaldtse** [comment]
The `https://bib.ietf.org/public/rfc/bibxml{n}/*.xml` path is now called the "legacy BibXML path" where its use will be / is discouraged. A decision was made to intentionally not show this path, in favour of the datatracker authenticated paths.

The tools team certainly recognises this use case. Perhaps @rjsparks could provide more information?

**Comment by strogonoff** [comment]
@ajeanmahoney:

To expand on Ronald’s comment, from my somewhat limited perspective as project developer…

(For the purposes of below, I use the term “xml2rfc-style path” to refer to paths that start with `/public/rfc/`. Also, links use dev.bibxml.org staging instance URLs, but production instance at bib.ietf.org should provide similar results.)

## Exporting bibliographic data as XML using xml2rfc-style path API


1. The service intentionally anchors to authoritative bibliographic data sources. They are “the database”, and they power service’s main GUI and API.

   (Xml2rfc tools XML datasets are not considered authoritative, as they were built by hand and sometimes contain mistakes. Examples of authoritative sources include Datatracker, RFC editor, rsync.iana.org, 3gpp.org, see [currently used sources](https://github.com/ietf-ribose/bibxml-service/wiki/Data-sources).)

2. To satisfy the RFP and make sure existing tooling still works, the service _does_ continue to serve the existing xml2rfc-style paths, and attempts to map them to authoritative bibliographic items.

   However, this map is one-way: it only attempts to resolve each xml2rfc path to an up-to-date bibliographic item obtained from authoritative sources. It was supposed to be enough, since any uses of xml2rfc paths were expected to arise from preexisting documents and all future uses would instead use the main API described in sibling section.

   Furthermore, in medium term it was expected that xml2rfc-style paths would be phased out in favor of primary API relying on more standardized document identifiers (as opposed to hand-crafted filenames of xml2rfc paths).

## Exporting bibliographic data as XML using service’s main API

Note that, if you log in via Datatracker, each bibliographic item details page contains a bibxml export link on the top. That link leads to service’s main API [endpoint for bibliographic item export](https://dev.bibxml.org/api/v1/#operation/getBibItemByDocId). It returns XML compliant with RFC 7991/xml2rfc spec.

There are two possible issues with using this method:

- Service’s main API requires authentication. (Note that if you try to access the same XML API URL without having logged in via Datatracker, the endpoint will report an error.) The tooling that processes that `<xi:include>` needs to support [Datatracker token based authentication](https://dev.bibxml.org/static/docs/topics/auth.html#http-header-token).

- Concerning Internet Drafts in particular, currently service’s main API requires to specify a version explicitly.
  
  - API endpoint for an I-D version ([like `draft-ietf-tcpm-rfc793bis-28`](https://dev.bibxml.org/get-one/by-docid/?docid=draft-ietf-tcpm-rfc793bis-28&doctype=Internet-Draft&query=%7B%22docid%22%3A%20%5B%7B%22id%22%3A%20%22draft-ietf-tcpm-rfc793bis-28%22%7D%5D%7D&query_format=json_struct&page=1)) returns a `<reference>` for that I-D version. 
  - API endpoint for “umbrella” I-D bibitem ([like `ietf-tcpm-rfc793bis`](https://dev.bibxml.org/get-one/by-docid/?docid=draft-ietf-tcpm-rfc793bis&doctype=Internet-Draft&query=ietf-tcpm-rfc793bis&query_format=docid_regex&page=1)) returns XML with `<referencegroup>` containing a `<reference>` for each version of that I-D (similar to an RFC subseries entry).


## Summary

In short, right now the service is not equipped to answer the question “which xml2rfc path corresponds to a bibliographic item shown in GUI”. It can be changed to capture this data, it would violate some principles on which it’s based (conceptually, xml2rfc path is not part of bibliographic data) but it’s not very effort-intensive. However, we should see if it’s truly required, since the primary XML API might cover the use case described in this ticket.

**Comment by rjsparks** [comment]
@ronaldtse @strogonoff We have talked past each other.

We _CANNOT_ deprecate/discourage the use of "legacy" paths.

There may be a future where we can change the RFCXML grammar to use tokens that could be used with the new api instead of providing paths for xi-includes, but until that day, the _majority_ of use will be documents that use xi-include with a path that looks like one of the legacy paths.

For now, they must be easy to find, and will be the _primary_ bit of data people are interested in.

_IN PARTICULAR_ people will want the path to the versionless internet draft data.

I really thought we had gotten onto the same page with this before. If we need a call to finalize this, let me know.

**Comment by strogonoff** [comment]
Our assumption that IETF will shortly start migrating away from xml2rfc paths may have been a misinterpretation of the RFP. xml2rfc adapter layer (point 2 in my comment above) was expanded to support reversing /public/rfc paths now:

<img width="1222" alt="Screenshot 2022-07-03 at 12 01 58 AM" src="https://user-images.githubusercontent.com/292959/177009956-45f54beb-3064-46e0-a0cc-60e75362c113.png">

**Comment by ronaldtse** [comment]
This has been resolved via email communication and feature deployed. Closing.

**Comment by ajeanmahoney** [comment]
@rjsparks I would like to reopen this issue.

The webpage shows the following:

    http://localhost:8003/public/rfc/bibxml/reference.RFC.8446.xml

rather than

    https://bib.ietf.org/public/rfc/bibxml/reference.RFC.8446.xml

Please see https://bib.ietf.org/get-bibliographic-item/?query=RFC8446, for instance.

**Comment by strogonoff** [comment]
@kesara I think there may have been a problem with production configuration.

See how Django obtains the value of request: https://github.com/django/django/blob/a4cb1dae6362d473116bb8c8fd97df89bfe036f8/django/http/request.py#L108-L124

- Are we sure your front-end proxies/CDN set `HTTP_HOST` or `SERVER_NAME` request headers? This service makes them unnecessary in some contexts, but Django may still rely on them.
- Or do they set `HTTP_X_FORWARDED_FOR`? In that case, adding a setting may help.
- By the way, if `localhost` successfully obtained it could mean you have `localhost` in `ALLOWED_HOSTS`, which should normally be avoided…

Nevertheless, I changed how URLs are reversed, so let’s see how it works in version `2022.8.3_1`…

**Comment by kesara** [comment]
@strogonoff Setting the `X-Forwarded-For` and `HOSTNAME` seemed to solve the issue.
I did include the `localhost` in the `ALLOWED_HOSTS` due to an issue in an earlier stage, which I can't remember now.
I have now removed that from the `ALLOWED_HOSTS` and seems to work well.
I'll deploy `2022.8.3_4` tomorrow.
Thanks.

**Comment by kesara** [comment]
@strogonoff I have deployed `2022.8.3_4` on bib.ietf.org.
Now the URL doesn't have the domain name.

```
//bib.ietf.org/public/rfc/bibxml/reference.RFC.9000.xml
```

Example: https://bib.ietf.org/get-bibliographic-item/?query=RFC+9000

**Comment by ronaldtse** [comment]
This is now fixed in #257 (issue described in #253) .

<img width="863" alt="Screen Shot 2022-08-08 at 3 29 11 PM" src="https://user-images.githubusercontent.com/11865/183363506-929fce36-0679-4ea2-b4d1-6cc6946216b8.png">

@ajeanmahoney could you let us know if this suffices for the purpose in this issue? If so please feel free to close this. Thanks!

**Comment by ajeanmahoney** [comment]
The URLs look good. Thanks!
